### PR TITLE
MAINT,BUG: Never import distutils above 3.12 [f2py]

### LINKED ISSUE #24838: BUG: `f2py` still attempts to import `numpy.distutils` on Python 3.12 when `--fcompiler` option is specified.
### Describe the issue:
`f2py` implements some logic to use the meson backend automatically on Python 3.12, where `numpy.distutils` no longer exists.

However, when calling `f2py` with the `--fcompiler` option, `numpy.distutils` is still imported, which raises `ImportError` when using running `f2py` with Python 3.12. This is due to the logic in these lines: https://github.com/numpy/numpy/blob/main/numpy/f2py/f2py2e.py#L608-L610

If the `--fcompiler` is not supported with the meson backend, perhaps a more useful error message could be thrown earlier?

### Reproduce the code example:

```python
f2py --fcompiler=gfortran -c -m
```

### Error message:

```shell
Cannot use distutils backend with Python 3.12, using meson backend instead.Traceback (most recent call last):
  File "/Users/phil/miniforge3/envs/meson/bin/f2py", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/Users/phil/miniforge3/envs/meson/lib/python3.12/site-packages/numpy/f2py/f2py2e.py", line 732, in main
    run_compile()
  File "/Users/phil/miniforge3/envs/meson/lib/python3.12/site-packages/numpy/f2py/f2py2e.py", line 610, in run_compile
    from numpy.distutils import fcompiler
ModuleNotFoundError: No module named 'numpy.distutils'
```

### Runtime information:

```
>>> import sys, numpy; print(numpy.__version__); print(sys.version); print(numpy.show_runtime())
1.26.0
3.12.0rc3 | packaged by conda-forge | (main, Sep 26 2023, 22:14:47) [Clang 15.0.7 ]
WARNING: `threadpoolctl` not found in system! Install it by `pip install threadpoolctl`. Once installed, try `np.show_runtime` again for more detailed build information
[{'numpy_version': '1.26.0',
  'python': '3.12.0rc3 | packaged by conda-forge | (main, Sep 26 2023, '
            '22:14:47) [Clang 15.0.7 ]',
  'uname': uname_result(system='Darwin', node='Phillips-MacBook-Pro.local', release='22.5.0', version='Darwin Kernel Version 22.5.0: Thu Jun  8 22:22:20 PDT 2023; root:xnu-8796.121.3~7/RELEASE_ARM64_T6000', machine='x86_64')},
 {'simd_extensions': {'baseline': ['SSE', 'SSE2', 'SSE3', 'SSSE3'],
                      'found': ['SSE41', 'POPCNT', 'SSE42'],
                      'not_found': ['AVX',
                                    'F16C',
                                    'FMA3',
                                    'AVX2',
                                    'AVX512F',
                                    'AVX512CD',
                                    'AVX512_KNL',
                                    'AVX512_SKX',
                                    'AVX512_CLX',
                                    'AVX512_CNL',
                                    'AVX512_ICL']}}]
```

### Context for the issue:

_No response_

### PR DESCRIPTION
**Pull request by HaoZeke**
Closes gh-24838. There were two locations in `f2py` where the options passed in would trigger an import of `distutils`. This simply guards them via an `if`.

### CODE REVIEW

File: numpy/f2py/f2py2e.py [hunk 1]
```diff
@@ -607,21 +609,27 @@ def run_compile():
...
-                try:
-                    nv = vmap[ov]
-                except KeyError:
-                    if ov not in vmap.values():
-                        print('Unknown vendor: "%s"' % (s[len(v):]))
-                nv = ov
-            i = flib_flags.index(s)
-            flib_flags[i] = '--fcompiler=' + nv
-            continue
+            if sys.version_info >= (3, 12):
```
Comment (mattip) [inline numpy/f2py/f2py2e.py:612]:
```suggestion
            if MESON_ONLY_VER:
```

**Comment by HaoZeke** [comment]
I think this could do with a backport @charris :)

**Comment by mattip** [comment]
Is there a test that failed before/passes after this PR? Maybe one that calls `subprocess.run` with `f2py --fcompiler=gfortran -c -m` ?

File: numpy/f2py/f2py2e.py [hunk 2]
```diff
@@ -707,8 +715,11 @@ def run_compile():
 def main():
     if '--help-link' in sys.argv[1:]:
         sys.argv.remove('--help-link')
-        from numpy.distutils.system_info import show_all
-        show_all()
+        if not MESON_ONLY_VER:
```
Comment (charris) [inline numpy/f2py/f2py2e.py:718]: Might be nicer to omit the note and exchange blocks.

**Comment by HaoZeke** [comment]
> Is there a test that failed before/passes after this PR? Maybe one that calls `subprocess.run` with `f2py --fcompiler=gfortran -c -m` ?

Well, it wouldn't trigger on the CI until #25111 ..

**Comment by mattip** [comment]
After #25181, needs a rebase

**Comment by HaoZeke** [comment]
> After #25181, needs a rebase

Yup, done, though maybe merging #25111 first would be better? Then the tests would run more comprehensively.

**Comment by rgommers** [comment]
This has a conflict, can you rebase @HaoZeke?



> Yup, done, though maybe merging #25111 first would be better? Then the tests would run more comprehensively.

This is an actual bug fix and quite small (most of the diff is a code move, the actual fixes are easy to identify). I'm seeing `f2py` problems in CI, so I'd prefer to get this in.

**Review by rgommers — APPROVED** [review]
This LGTM, in it goes. Thanks @HaoZeke & reviewers.

**Pull request merged**